rust/blockstore/src/arrow/ordered_blockfile_writer.rs

```diff
@@ -0,0 +1,955 @@
+    remaining_block_stack: VecDeque<BlockIdAndEndKey>,
+    /// Holds the current block delta and its end key. When we receive a write past the end key, this delta is moved into `completed_block_deltas`.
+    current_block_delta: Option<CurrentDeltaAndEndKey>,
+    /// Deltas in this vec can no longer receive writes and are ready to be committed.
+    completed_block_deltas: Vec<OrderedMergeBlockDelta>,
+}
+
+#[derive(Clone)]
+pub struct ArrowOrderedBlockfileWriter {
+    block_manager: BlockManager,
+    root_manager: RootManager,
+    root: RootWriter,
+    inner: Arc<Mutex<Inner>>,
+    id: Uuid,
+}
+
+#[derive(Error, Debug)]
+pub enum ArrowBlockfileError {
+    #[error("Block not found")]
+    BlockNotFound,
+}
+
+impl ChromaError for ArrowBlockfileError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            ArrowBlockfileError::BlockNotFound => ErrorCodes::Internal,
+        }
+    }
+}
+
+impl ArrowOrderedBlockfileWriter {
+    pub(super) fn new<K: ArrowWriteableKey, V: ArrowWriteableValue>(
+        id: Uuid,
+        block_manager: BlockManager,
+        root_manager: RootManager,
+    ) -> Self {
+        let initial_block = block_manager.create::<K, V, OrderedMergeBlockDelta>();
+        let sparse_index = SparseIndexWriter::new(initial_block.id);
+        let root_writer = RootWriter::new(CURRENT_VERSION, id, sparse_index);
+
+        Self {
+            block_manager,
+            root_manager,
+            root: root_writer,
+            id,
+            inner: Arc::new(Mutex::new(Inner {
+                current_block_delta: Some((initial_block, None)),
+                completed_block_deltas: Vec::new(),
+                remaining_block_stack: VecDeque::new(),
+            })),
+        }
+    }
+
+    pub(super) fn from_root(
+        id: Uuid,
+        block_manager: BlockManager,
+        root_manager: RootManager,
+        new_root: RootWriter,
+    ) -> Self {
+        let remaining_block_stack = {
+            let root_forward = &new_root.sparse_index.data.lock().forward;
+
+            root_forward
+                .iter()
+                .zip_longest(root_forward.iter().skip(1))
+                .map(|zipped| match zipped {
+                    itertools::EitherOrBoth::Both(block, next_block) => match next_block.0 {
+                        SparseIndexDelimiter::Start => {
+                            // The next block should never be a start block
+                            panic!("Invariant violated: unexpected start delimiter.")
+                        }
+                        SparseIndexDelimiter::Key(end_key) => (*block.1, Some(end_key.clone())),
+                    },
+                    itertools::EitherOrBoth::Left(block) => (*block.1, None),
+                    itertools::EitherOrBoth::Right(_) => {
+                        unreachable!()
+                    }
+                })
+                .collect::<VecDeque<_>>()
+        };
```

Comment:
can we push this into the sparse index

Reply:
not easily because we need to take a lock on the sparse index data
if the method on the sparse index materialized into a `Vec` and returned that it would work, but then we would need to remap into a `VecDeque`